packages/web-api/src/geolocation.ts

```diff
@@ -0,0 +1,271 @@
+    successCallback: (position: CustomGeolocationPosition) => void,
+    errorCallback: (error: CustomGeolocationError) => void
+  ) => Unsubscribe;
+};
+
+type Geolocation = {
+  $location: Store<{ latitude: number; longitude: number } | null>;
+  $latitude: Store<number | null>;
+  $longitude: Store<number | null>;
+  request: EventCallable<void>;
+  watching: {
+    start: EventCallable<void>;
+    stop: EventCallable<void>;
+    $active: Store<boolean>;
+  };
+  reporting: {
+    failed: Event<CustomGeolocationError | globalThis.GeolocationPositionError>;
+  };
+};
+
+const BrowserProvider = Symbol('BrowserProvider');
+
+export function trackGeolocation(
+  params?: GeolocationParams & {
+    providers?: Array<
+      typeof BrowserProvider | CustomProvider | globalThis.Geolocation
+    >;
+  }
+): Geolocation {
+  const providres = (
+    params?.providers ?? /* In case of no providers, we will use the default one only */ [
+      BrowserProvider,
+    ]
+  )
+    .map((provider) => {
+      /* BrowserProvider symbol means usage of navigator.geolocation */
+      if (provider === BrowserProvider) {
+        const browserGeolocationAvailable =
+          globalThis.navigator && 'geolocation' in globalThis.navigator;
+        if (!browserGeolocationAvailable) {
+          return null;
+        }
+
+        return globalThis.navigator.geolocation;
+      }
+
+      if (isDefaultProvider(provider)) {
+        return provider;
+      }
+
+      return provider(params ?? {});
+    })
+    .filter(Boolean) as Array<
+    ReturnType<CustomProvider> | globalThis.Geolocation
+  >;
+
+  // -- units
+
+  const $location = createStore<{
+    latitude: number;
+    longitude: number;
+  } | null>(null);
+
+  const $longitude = combine(
+    $location,
+    (location) => location?.longitude ?? null
+  );
+
+  const $latitude = combine(
+    $location,
+    (location) => location?.latitude ?? null
+  );
+
+  const request = createEvent();
+
+  const startWatching = createEvent();
+  const stopWatching = createEvent();
+  const $watchingActive = createStore(false);
+
+  const failed = createEvent<
+    CustomGeolocationError | globalThis.GeolocationPositionError
+  >();
+
+  // -- shared logic
+
+  const newPosition = createEvent<
+    CustomGeolocationPosition | globalThis.GeolocationPosition
+  >();
+
+  sample({
+    clock: newPosition,
+    fn: (r) => ({ latitude: r.coords.latitude, longitude: r.coords.longitude }),
+    target: $location,
+  });
+
+  // -- get current position
+
+  const getCurrentPositionFx = createEffect<
+    void,
+    CustomGeolocationPosition | globalThis.GeolocationPosition,
+    CustomGeolocationError | globalThis.GeolocationPositionError
+  >(async () => {
+    let geolocation:
+      | globalThis.GeolocationPosition
+      | CustomGeolocationPosition
+      | null = null;
+
+    for (const provider of providres) {
+      if (isDefaultProvider(provider)) {
+        geolocation = await new Promise<GeolocationPosition>(
```

Comment:
Actually, looks like there must be a safe-call wrapper - if i'm not missing anything, now, if first provider is failed - any other is not called at all